Bluetooth: gatt: Add support for automatic ATT MTU Exchange

MariuszSkamra opened the pull request:
This adds support for automatic ATT MTU Exchange that will be done right after the connection has been established.
This ensures the ATT MTU Exchange request will not be sent more than once. As the improvement the code follows Core 5.3 recommendation to skip MTU Exchange procedure if already performed by peer.

Fixes: #43946

MariuszSkamra committed:
Bluetooth: gatt: Add support for automatic ATT MTU Exchange

This adds support for automatic ATT MTU Exchange that will be done right
after the connection has been established.

Fixes: #43946
Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: gatt: Disallow repeated attempts to exchange MTU

The MTU can be exchange once during the connection by the client.
This ensures the ATT MTU Exchange request will not be sent again.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: gatt: Skip MTU exchange if already initiated by peer

This implements Core 5.3 recommendation to skip MTU Exchange procedure
if already performed by peer.

Core 5.3 | Vol 3, Part F 3.4.2.2:
If MTU is exchanged in one direction, that is sufficient for both
directions.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: audio: Enable auto ATT MTU Exchange for BASS and unicast client

This enables automatic MTU exchange for services that require MTU size
bigger than default ATT MTU to operate.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5158,7 +5170,9 @@ void bt_gatt_connected(struct bt_conn *conn)
 #if defined(CONFIG_BT_GATT_CLIENT)
 	add_subscriptions(conn);
-
+#if defined(CONFIG_BT_GATT_AUTO_UPDATE_MTU)
+	bt_gatt_exchange_mtu(conn, &gatt_exchange_params);
+#endif /* CONFIG_BT_GATT_AUTO_UPDATE_MTU */
```
Do we need to check the role? 

The core spec says:

> The Client Rx MTU shall be greater than or equal to the default ATT_MTU.
> This request shall only be sent once during a connection by the client.

But who is the client?

For e.g. LE Audio both sides will act as a GATT server and GATT client, so both sides will attempt to do `bt_gatt_exchange_mtu` at the same time. We could consider adding a small delay for e.g. the peripheral. The same discussion was done for auto-connect EATT: https://github.com/zephyrproject-rtos/zephyr/pull/43112#pullrequestreview-902359990

MariuszSkamra replied:
The both sides can be Client and Server at the same time. Even if both will initiate MTU Exchange at the same time there will no be a collision since the Tx and Rx MTU is the same (symmetrical).
Will take a look at the delay discossion :)

MariuszSkamra replied:
@Thalley The discussion regarding EATT auto-connection is somehow different, as there might be collision when device allocated EATT Channels and sent request. When it received EATT connection request it check whether it has free channels, but it does not have because those might be in use (self-initiated request in progress) and return error. This will result in EATT Channels not being created.

Here, regarding MTU, there will not be collision, because MTU in ATT MTU Request has to be the same as the one that would be sent in ATT MTU Response.

Thalley replied:
@MariuszSkamra But if both sides requests MTU exchange on connect (e.g. on 2 Zephyr devices), then both of them will fail, right?

But I see that we accept MTU requests even if `ATT_MTU_EXCHANGE_COMPLETE` or `ATT_MTU_EXCHANGE_IN_PROGRESS` is set, so it will currently not be a problem. 

If we have sent an MTU exchange to the peer, should we accept MTU exchange requests from that device too? i.e.

Central sends MTU request (sets `ATT_MTU_EXCHANGE_IN_PROGRESS`)
Peripheral sends MTU request (sets `ATT_MTU_EXCHANGE_IN_PROGRESS`)
Peripheral receives central's MTU request => Accept or reject since we are already waiting response?
Central receives peripheral's MTU request => Accept or reject since we are already waiting response?


There's also the other issue regarding multiple requests:
The spec states that a client shall only request MTU once. Should we reject multiple MTU requests from a client as a GATT server?

MariuszSkamra replied:
> @MariuszSkamra But if both sides requests MTU exchange on connect (e.g. on 2 Zephyr devices), will both accept or reject?
Both accept. Why would Client reject MTU Exchange?
> 
> But I see that we accept MTU requests even if `ATT_MTU_EXCHANGE_COMPLETE` or `ATT_MTU_EXCHANGE_IN_PROGRESS` is set, so it will currently not be a problem.
> 
> If we have sent an MTU exchange to the peer, should we accept MTU exchange requests from that device too? i.e.
> 
> Central sends MTU request (sets `ATT_MTU_EXCHANGE_IN_PROGRESS`) Peripheral sends MTU request (sets `ATT_MTU_EXCHANGE_IN_PROGRESS`) Peripheral receives central's MTU request => Accept or reject since we are already waiting response? Central receives peripheral's MTU request => Accept or reject since we are already waiting response?
As the both sides can request MTU Exchange, the request shall not be rejected. The `ATT_MTU_EXCHANGE_IN_PROGRESS` and `ATT_MTU_EXCHANGE_COMPLETE` have been added to ensure the Client will not request MTU Exchange more than once. 
> 
> There's also the other issue regarding multiple requests: The spec states that a client shall only request MTU once. Should we reject multiple MTU requests from a client as a GATT server?

This would lead to potential interoperability issues, so I think it's better to sent the ATT MTU Response. We should be compliant with the Core but have a little sense of flexibility to what peer does.

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5158,7 +5170,9 @@ void bt_gatt_connected(struct bt_conn *conn)
 	add_subscriptions(conn);
-
+#if defined(CONFIG_BT_GATT_AUTO_UPDATE_MTU)
+	bt_gatt_exchange_mtu(conn, &gatt_exchange_params);
```
Since `bt_gatt_exchange_mtu` can fail we should either handle that or at the very least log it.

MariuszSkamra commented:
Changes made:

- Excluded `ATT_MTU_EXCHANGE_*` flags when `CONFIG_BT_GATT_CLIENT` is not set
- Added warning when auto MTU Exchange failed

Thalley commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -4896,6 +4896,18 @@ static void add_subscriptions(struct bt_conn *conn)
 }
+
+#if defined(CONFIG_BT_GATT_AUTO_UPDATE_MTU)
+static void gatt_exchange_mtu_func(struct bt_conn *conn, uint8_t err,
+				   struct bt_gatt_exchange_params *params)
+{
+	BT_DBG("conn %p err 0x%02x", conn, err);
+}
```
Should we similarly have a BT_WARN here if `err != 0`?

alwa-nordic commented on subsys/bluetooth/host/att.c:
```diff
@@ -3310,6 +3317,25 @@ int bt_att_req_send(struct bt_conn *conn, struct bt_att_req *req)
 		return -ENOTCONN;
 	}
 
+#if defined(CONFIG_BT_GATT_CLIENT)
+	struct bt_att_hdr *hdr;
+
+	hdr = (void *)req->buf->data;
+
+	if (hdr->code == BT_ATT_OP_MTU_REQ) {
+		struct bt_att_chan *att_chan = att_get_fixed_chan(conn);
+
+		/* This request shall only be sent once during a connection by the client. */
+		if (atomic_test_bit(att_chan->flags, ATT_MTU_EXCHANGE_COMPLETE)) {
+			return -ECANCELED;
+		}
+
+		if (atomic_test_and_set_bit(att_chan->flags, ATT_MTU_EXCHANGE_IN_PROGRESS)) {
+			return -EALREADY;
+		}
+	}
+#endif /* CONFIG_BT_GATT_CLIENT */
+
```
I'm a bit uncomfortable with special handling for a specific opcode in `bt_att_req_send`. It feel a bit arbitrarily placed here. I'm not sure enforcing the legality of sending a request should be the responsibility of `bt_att_req_send`.

MariuszSkamra replied:
You have right, this is not a proper place for it. I will change that

MariuszSkamra commented on subsys/bluetooth/host/att.c:
```diff
@@ -3310,6 +3317,25 @@ int bt_att_req_send(struct bt_conn *conn, struct bt_att_req *req)
+			return -ECANCELED;
+		}
+
+		if (atomic_test_and_set_bit(att_chan->flags, ATT_MTU_EXCHANGE_IN_PROGRESS)) {
```
Potential ATT Error Response is not handled, leaving ATT_MTU_EXCHANGE_IN_PROGRESS flag set

alwa-nordic commented on subsys/bluetooth/host/Kconfig.gatt:
```diff
@@ -141,6 +141,12 @@ config BT_GATT_AUTO_DISCOVER_CCC
 	  handles if the CCC handle is unknown by the application.
 
+config BT_GATT_AUTO_UPDATE_MTU
+	bool "Support to automatic MTU update"
```
- "Support to" sounds ungrammatical. It should be just "Support".
- I would like to drop "Support" because it makes the sentence unnecessarily vague.
- We should specify which MTU value we refer to, like "ATT MTU".

```suggestion
	bool "Automatically send the ATT MTU exchange request on connect"
```

asbjornsabo replied:
I would even say `bool "Automatically send ATT MTU exchange request on connect"` (drop "the").

MariuszSkamra commented on subsys/bluetooth/host/att.c:
```diff
@@ -675,6 +680,8 @@ static uint8_t att_mtu_rsp(struct bt_att_chan *chan, struct net_buf *buf)
 	 */
 	chan->chan.tx.mtu = chan->chan.rx.mtu;
 
+	atomic_set_bit(chan->flags, ATT_MTU_EXCHANGE_COMPLETE);
```
Could, however the `ATT_MTU_EXCHANGE_COMPLETE` takes the precedence so `ATT_MTU_EXCHANGE_IN_PROGRESS` is not checked

alwa-nordic replied:
Leaving the flag raised while calling it "in progress" is bad. Either clear the flag or give it a name that says the process may be already complete. E.g. "initiated", "started"..

MariuszSkamra replied:
BTW, I realized I do not even need the `ATT_MTU_EXCHANGE_IN_PROGRESS`

Thalley replied:
> BTW, I realized I do not even need the `ATT_MTU_EXCHANGE_IN_PROGRESS`

Oh? How do we ensure that `bt_gatt_exchange_mtu` is not rejected on the second call then?

MariuszSkamra replied:
The Core spec says:
"This request shall only be sent once during a connection by the client."
So the client cannot sent it again, so it does not matter if it's in progress or already completed. If the client sends the request it sets the flag. Even if the request was rejected it cannot sent it again

Thalley replied:
Ah, right. So it's not the procedure that cannot be done more than once; the request cannot be send again, even if it is for some reason rejected by the server. In that case then yes, if we set `ATT_MTU_EXCHANGE_COMPLETE` once the request has been sent, it should be fine.

alwa-nordic commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -5158,7 +5170,14 @@ void bt_gatt_connected(struct bt_conn *conn)
 
 #if defined(CONFIG_BT_GATT_CLIENT)
 	add_subscriptions(conn);
+#if defined(CONFIG_BT_GATT_AUTO_UPDATE_MTU)
+	int err;
 
+	err = bt_gatt_exchange_mtu(conn, &gatt_exchange_params);
+	if (err) {
+		BT_WARN("MTU Exchange failed (err %d)", err);
+	}
+#endif /* CONFIG_BT_GATT_AUTO_UPDATE_MTU */
 #endif /* CONFIG_BT_GATT_CLIENT */
```
- Subscription upkeep and MTU exchange are unrelated. We should keep it separate in code as well.
- It's simpler if enabling automatic MTU exchange depends only on `CONFIG_BT_GATT_AUTO_UPDATE_MTU`. That is what that kconfig's name means to me.
- It's more correct if this fails to compile if `depends on BT_GATT_CLIENT` is removed from `CONFIG_BT_GATT_AUTO_UPDATE_MTU` but `bt_gatt_exchange_mtu` is not defined instead of silently ignoring `CONFIG_BT_GATT_AUTO_UPDATE_MTU`.

```suggestion
#endif /* CONFIG_BT_GATT_CLIENT */

#if defined(CONFIG_BT_GATT_AUTO_UPDATE_MTU)
	int err;

	err = bt_gatt_exchange_mtu(conn, &gatt_exchange_params);
	if (err) {
		BT_WARN("MTU Exchange failed (err %d)", err);
	}
#endif /* CONFIG_BT_GATT_AUTO_UPDATE_MTU */
```

MariuszSkamra committed:
Bluetooth: shell: Adapt to the ATT MTU API changes

This handles -EALREADY return error.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

MariuszSkamra committed:
Bluetooth: tests: Remove ATT MTU Exchange from VCS and MICS bsim tests

This patch removes the redundant test case step.
Neither VCS/VCP nor MICS/MICP require ATT MTU to be other than default
23 bytes.

Signed-off-by: Mariusz Skamra <mariusz.skamra@codecoup.pl>

Thalley commented on include/bluetooth/gatt.h:
```diff
@@ -1231,6 +1231,7 @@ struct bt_gatt_exchange_params {
  *  Allow a pending request to resolve before retrying, or call this function
  *  outside the BT RX thread to get blocking behavior. Queue size is controlled
  *  by @kconfig{CONFIG_BT_L2CAP_TX_BUF_COUNT}.
+ *  @retval -EALREADY The MTU exchange procedure has been already performed.
```
```suggestion
 *  by @kconfig{CONFIG_BT_L2CAP_TX_BUF_COUNT}.
 *
 *  @retval -EALREADY The MTU exchange procedure has been already performed.
```

MariuszSkamra replied:
done

Thalley reviewed (changes requested):
Minor comment regarding documentation, otherwise LGTM. 

Nice that you also updated the samples and BSIM tests :)
